deepspeed/utils/tensor_fragment.py

```diff
@@ -58,9 +62,95 @@ def get_full_hp_param(self, optim_state_key=None):
     return reduce_buffer.reshape_as(self)
 
 
+def get_full_hp_grad(self):
+    reduce_buffer = torch.zeros_like(self, dtype=torch.float32).flatten()
+    if self._hp_mapping is not None:
+        lp_frag_address = self._hp_mapping.lp_fragment_address
+        reduce_fragment = torch.narrow(reduce_buffer,
+                                       0,
+                                       lp_frag_address.start,
+                                       lp_frag_address.numel)
+
+        hp_mapping = self._hp_mapping
+
+        if hp_mapping.use_offload:
+            gradient_dict = hp_mapping.offload_gradient_dict
+        else:
+            gradient_dict = hp_mapping.gradient_dict
+
+        if hp_mapping.param_group_index not in gradient_dict or gradient_dict[
+                hp_mapping.param_group_index] is None:
+            raise ValueError(
+                "Gradients are only available immediately after backward and before engine step"
+            )
+
+        lp_grad_fragment = gradient_dict[hp_mapping.param_group_index][
+            self._index_in_param_group]
+        hp_grad_fragment = lp_grad_fragment.to(torch.float32).flatten()
+        reduce_fragment.data.copy_(hp_grad_fragment.data)
+    dist.all_reduce(reduce_buffer, group=self._dp_group)
+    return reduce_buffer.reshape_as(self)
+
+
+def safe_get_full_fp32_param(param):
+    """Assemble and return the fp32 parameter of a low-precision (e.g., fp16) parameter.
+
+        Args:
+            param (``torch.nn.Parameter``): The low-precision parameter
+    """
+    # ZeRO stage 3 param
+    if hasattr(param, 'ds_id'):
+        return param._z3_optimizer.get_full_hp_param(param)
+
+    # ZeRO stage 1, 2, and bf16_optimizer params
+    if hasattr(param, '_hp_mapping'):
+        return param.get_full_hp_param()
+    return None
+
+
+def safe_get_full_optimizer_state(param, optim_state_key):
+    """Assemble and return the fp32 optimizer state of a low-precision (e.g., fp16) parameter.
+
+        Args:
+            param (``torch.nn.Parameter``): The low-precision parameter
+    """
+    # ZeRO stage 3 param
+    if hasattr(param, 'ds_id'):
+        return param._z3_optimizer.get_full_hp_param(param, optim_state_key)
+
+    # ZeRO stage 1, 2, and bf16_optimizer params
+    if hasattr(param, '_hp_mapping'):
+        return param.get_full_hp_param(optim_state_key)
+    return None
+
+
+# TODO: Figure out the correct return dtype
+def safe_get_full_grad(param):
+    """Assemble and return the fp32 gradient of a low-precision (e.g., fp16) parameter.
+
+        Args:
+            param (``torch.nn.Parameter``): The low-precision parameter
+    """
+    if param.grad is not None:
+        return param.grad
+
+    # ZeRO stage 3 param
+    if hasattr(param, 'ds_id'):
+        return param._z3_optimizer.get_fp32_grad_for_param(param)
+
+    # ZeRO stage 1, 2, and bf16_optimizer params
+    if hasattr(param, '_hp_mapping'):
+        return param.get_full_hp_grad()
+    return None
```

Comment:
```suggestion
    # ZeRO stage 3 param
    if hasattr(param, 'ds_id'):
        return param._z3_optimizer.get_fp32_grad_for_param(param)

    # ZeRO stage 1, 2, and bf16_optimizer params
    if hasattr(param, '_hp_mapping'):
        return param.get_full_hp_grad()

    if param.grad is not None:
        return param.grad

        return None
```

this solves the hanging. 

no idea why for `size[1]` params, on many gpus, some of them have `param.grad is not None`